File: api/src/main/java/org/egglog/api/user/model/dto/response/UserResponse.java
package org.egglog.api.user.model.dto.response;

import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import jakarta.persistence.Column;
import jakarta.persistence.Temporal;
import jakarta.persistence.TemporalType;
import lombok.*;
import org.egglog.api.hospital.model.dto.response.HospitalAuthResponse;
import org.egglog.api.hospital.model.dto.response.UserHospitalResponse;
import org.egglog.api.user.model.entity.enums.UserRole;
import org.egglog.api.user.model.entity.enums.UserStatus;

import java.time.LocalDateTime;


/**
 * @packageName    : org.egglog.api.user.model.dto.response
 * @fileName       : UserResponse
 * @author         : 김형민
 * @date           : 2024-04-22
 * @description    : 공개할 수 있는 유저 정보 객체입니다.
 * ===========================================================
 * DATE              AUTHOR             NOTE
 * -----------------------------------------------------------
 * 2024-04-22        김형민       최초 생성
 */
@Builder
@Getter @Setter
@AllArgsConstructor
@NoArgsConstructor
public class UserResponse {
    private Long id;
    private String email;
    private String userName;
    private UserHospitalResponse selectedHospital;
    private HospitalAuthResponse hospitalAuth;
    private String profileImgUrl;
    private UserRole userRole;
    private UserStatus userStatus;
    private String deviceToken;
    private String empNo;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private LocalDateTime loginAt;
}

File: client/data/src/main/java/com/org/egglog/data/auth/model/UserResponse.kt
package com.org.egglog.data.auth.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.org.egglog.data.util.LocalDateTimeSerializer
import com.org.egglog.domain.auth.model.UserDetail
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import java.time.LocalDateTime

@Entity
@Serializable
data class UserResponse(
    @PrimaryKey val id: Long,
    val email: String,
    val userName: String,
    val profileImgUrl: String,
    val userRole: UserRole,
    val userStatus: UserStatus,
    @Contextual val selectedHospital: UserHospitalResponse?,
    @Contextual val hospitalAuth: HospitalAuthResponse?,
    val deviceToken: String?,
    val empNo: String?,
    @Serializable(with = LocalDateTimeSerializer::class) val createdAt: LocalDateTime?,
    @Serializable(with = LocalDateTimeSerializer::class) val updatedAt: LocalDateTime?,
    @Serializable(with = LocalDateTimeSerializer::class) val loginAt: LocalDateTime?,
    val workGroupId: Long?
)

fun UserResponse.toDomainModel(): UserDetail {
    return UserDetail(
        id = id,
        email = email,
        userName = userName,
        selectedHospital = selectedHospital?.toDomainModel(),
        hospitalAuth = hospitalAuth?.toDomainModel(),
        profileImgUrl = profileImgUrl,
        userRole = userRole.toString(),
        userStatus = userStatus.toString(),
        deviceToken = deviceToken,
        empNo = empNo,
        createdAt = createdAt,
        updatedAt = updatedAt,
        loginAt = loginAt,
        workGroupId = workGroupId
    )
}